### PR TITLE
fix: potential crash caused by dlopen different gtk libraries

### PR DESCRIPTION
**Pull request by deepak1556**
#### Description of Change

https://github.com/electron/electron/pull/19159 refactored file dialog implementation that relies requiring gtk library at runtime, around the same time chromium has been refactoring itself to also require gtk at runtime https://bugs.chromium.org/p/chromium/issues/detail?id=1192861

Since chromium already loads the required gtk library into the process on [startup](https://source.chromium.org/chromium/chromium/src/+/main:ui/gtk/gtk_compat.cc;l=137-140), the file dialog implementation should just get a handle to the loaded library and perform operations on it, the current implementation can lead to situations in classic snaps where two different version gtk libraries can get loaded into the same process triggering a crash, refs https://github.com/microsoft/vscode/issues/146584#issuecomment-1090490785

*NB:* There is no functionality change in this PR, simplifies library loading following upstream changes and fixes potential crash alongside.

/cc @tristan957

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [ ] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: fix crash when opening gtk file dialogs due to mismatched versions

### CODE REVIEW

**Comment by tristan957** [comment]
Hi @deepak1556. Thanks for tracking down the source of the issue. With your PR, I am a little confused how any of the `gtk_file_chooser_native_*` methods work because it is possible that those symbols don't exist.

Where is the definition of `electron::IsElectron_gtkInitialized()`?

**Comment by deepak1556** [comment]
`electron_gtk_stubs.cc` is responsible for loading the symbols and providing utilities to check if the required symbols are loaded, this file is generated during compile time. An example can be seen with upstream's [stub file](https://source.chromium.org/chromium/chromium/src/+/main:out/Debug/gen/ui/gtk/gtk_stubs.cc;l=2203) that gets generated for https://source.chromium.org/chromium/chromium/src/+/main:ui/gtk/gtk.sig, if the symbol is not present it will just call the utility function ([example](https://source.chromium.org/chromium/chromium/src/+/main:out/Debug/gen/ui/gtk/gtk_stubs.cc;l=2208-2209) is provided as part of `logging_function` in the `Build.gn` file. In our case, we provide a noop logger which allows for symbol load failures.

**Comment by tristan957** [comment]
Ok cool. Thanks for the background. This PR sounds like a good solution then. I wonder if Flatpak has the same issue.

File: shell/browser/ui/electron_gtk.sigs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+GtkFileChooserNative* gtk_file_chooser_native_new(const gchar* title, GtkWindow* parent, GtkFileChooserAction action, const gchar* accept_label, const gchar* cancel_label);
+void gtk_native_dialog_set_modal(GtkNativeDialog* self, gboolean modal);
+void gtk_native_dialog_show(GtkNativeDialog* self);
+void gtk_native_dialog_hide(GtkNativeDialog* self);
+gint gtk_native_dialog_run(GtkNativeDialog* self);
+void gtk_native_dialog_destroy(GtkNativeDialog* self);
```
Comment (nornagon) [inline shell/browser/ui/electron_gtk.sigs:6]: how was this file generated? Might we have to update it in future? Perhaps a comment in the GN file would be in order?

File: patches/chromium/make_gtk_getlibgtk_public.patch [hunk 2]
```diff
@@ -0,0 +1,52 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: deepak1556 <hop2deep@gmail.com>
+Date: Thu, 7 Apr 2022 20:30:16 +0900
+Subject: Make gtk::GetLibGtk public
+
+Allows embedders to get a handle to the gtk library
+already loaded in the process.
```
Comment (nornagon) [inline patches/chromium/make_gtk_getlibgtk_public.patch:7]: should this be upstreamed?

File: shell/browser/electron_browser_main_parts.cc [hunk 3]
```diff
@@ -367,6 +369,12 @@ void ElectronBrowserMainParts::ToolkitInitialized() {
   linux_ui->Initialize();
   DCHECK(ui::LinuxInputMethodContextFactory::instance());
 
+  // Try loading gtk symbols used by Electron.
+  electron::InitializeElectron_gtk(gtk::GetLibGtk());
+  if (!electron::IsElectron_gtkInitialized()) {
+    electron::UninitializeElectron_gtk();
```
Comment (nornagon) [inline shell/browser/electron_browser_main_parts.cc:375]: this is somewhat confusing logic. what does it mean for gtk initialization to have failed but we continue running the app? Is this error fatal?

**Review by nornagon — APPROVED** [review]

File: shell/browser/ui/electron_gtk.sigs (hunk 1 above)
Comment (deepak1556) [reply shell/browser/ui/electron_gtk.sigs:6]: Signatures are manually declared here for which need function ptrs, currently it covers the optional functions for the native file chooser implementation https://github.com/electron/electron/blob/4d4682c0e38cd49add16310278726e48de9db2f5/shell/browser/ui/file_dialog_gtk.cc#L29-L38. But moving forward to support gtk4 and gtk3 at runtime, this file will have to be extended for all gtk functions we use and then perform necessary runtime checking before using the ptr.

The `.sig` file does not seem to take comment section, I will add this info to the GN file.

File: patches/chromium/make_gtk_getlibgtk_public.patch (hunk 2 above)
Comment (deepak1556) [reply patches/chromium/make_gtk_getlibgtk_public.patch:7]: I can give it an attempt.

**Review by zcbenz — APPROVED** [review]

File: shell/browser/ui/electron_gtk.sigs (hunk 1 above)
Comment (deepak1556) [reply shell/browser/ui/electron_gtk.sigs:6]: Addressed in https://github.com/electron/electron/pull/33818/commits/6c9d92c41d382ba627b803ede5d7b7a153bebbb5

**Pull request merged**